File: src/sdk/nodejs/ofa-agent/src/connection.ts
/**
 * Connection Module
 * Sprint 29: Node.js Agent SDK
 */

import WebSocket from 'ws';
import axios from 'axios';
import { AgentConfig } from './config';
import { Message } from './message';

/**
 * 连接接口
 */
export interface Connection {
  connect(): Promise<void>;
  disconnect(): Promise<void>;
  send(msg: Message): Promise<void>;
  receive(): Promise<Message | null>;
  isConnected(): boolean;
}

/**
 * 创建连接
 */
export function createConnection(config: AgentConfig): Connection {
  switch (config.connectionType) {
    case 'http':
      return new HttpConnection(config);
    case 'websocket':
      return new WebSocketConnection(config);
    case 'grpc':
      return new GrpcConnection(config);
    default:
      return new HttpConnection(config);
  }
}

/**
 * 连接类型
 */
export enum ConnectionType {
  HTTP = 'http',
  WEBSOCKET = 'websocket',
  GRPC = 'grpc',
}

/**
 * HTTP连接
 */
export class HttpConnection implements Connection {
  private config: AgentConfig;
  private connected: boolean = false;
  private baseUrl: string = '';
  private messageQueue: Message[] = [];

  constructor(config: AgentConfig) {
    this.config = config;
  }

  async connect(): Promise<void> {
    this.baseUrl = `http://${this.config.centerUrl}`;
    this.connected = true;

    // 注册Agent
    await axios.post(`${this.baseUrl}/api/v1/agents/register`, {
      id: this.config.agentId,
      name: this.config.name,
      type: 'nodejs',
    });

    // 启动消息轮询
    this.startPolling();
  }

  async disconnect(): Promise<void> {
    this.connected = false;
  }

  async send(msg: Message): Promise<void> {
    if (!this.connected) throw new Error('Not connected');

    const type = msg.type.toLowerCase();
    await axios.post(`${this.baseUrl}/api/v1/${type}`, msg);
  }

  async receive(): Promise<Message | null> {
    return this.messageQueue.shift() || null;
  }

  isConnected(): boolean {
    return this.connected;
  }

  private startPolling(): void {
    const poll = async () => {
      if (!this.connected) return;

      try {
        const response = await axios.get(
          `${this.baseUrl}/api/v1/agents/${this.config.agentId}/messages`
        );

        for (const msg of response.data) {
          this.messageQueue.push(msg);
        }
      } catch (error) {
        console.error('Polling error:', error);
      }

      setTimeout(poll, 1000);
    };

    poll();
  }
}

/**
 * WebSocket连接
 */
export class WebSocketConnection implements Connection {
  private config: AgentConfig;
  private ws: WebSocket | null = null;
  private connected: boolean = false;
  private messageQueue: Message[] = [];

  constructor(config: AgentConfig) {
    this.config = config;
  }

  async connect(): Promise<void> {
    const url = `ws://${this.config.centerUrl}/ws`;

    return new Promise((resolve, reject) => {
      this.ws = new WebSocket(url);

      this.ws.on('open', () => {
        this.connected = true;
        resolve();
      });

      this.ws.on('message', (data: Buffer) => {
        try {
          const msg = JSON.parse(data.toString()) as Message;
          this.messageQueue.push(msg);
        } catch (error) {
          console.error('Parse error:', error);
        }
      });

      this.ws.on('close', () => {
        this.connected = false;
      });

      this.ws.on('error', (error: Error) => {
        reject(error);
      });
    });
  }

  async disconnect(): Promise<void> {
    this.connected = false;
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  async send(msg: Message): Promise<void> {
    if (!this.ws || !this.connected) throw new Error('Not connected');
    this.ws.send(JSON.stringify(msg));
  }

  async receive(): Promise<Message | null> {
    return this.messageQueue.shift() || null;
  }

  isConnected(): boolean {
    return this.connected;
  }
}

/**
 * gRPC连接（模拟）
 */
export class GrpcConnection implements Connection {
  private config: AgentConfig;
  private connected: boolean = false;

  constructor(config: AgentConfig) {
    this.config = config;
  }

  async connect(): Promise<void> {
    // 实际实现需要@grpc/grpc-js
    this.connected = true;
  }

  async disconnect(): Promise<void> {
    this.connected = false;
  }

  async send(_msg: Message): Promise<void> {
    // gRPC发送
  }

  async receive(): Promise<Message | null> {
    return null;
  }

  isConnected(): boolean {
    return this.connected;
  }
}

File: src/sdk/python/ofa_agent/connection.py
"""
Connection Module
Sprint 29: Python Agent SDK
"""

import asyncio
import json
import logging
import aiohttp
import websockets
from typing import Any, Dict, Optional
from dataclasses import dataclass

logger = logging.getLogger(__name__)


class ConnectionType:
    """连接类型"""
    GRPC = "grpc"
    HTTP = "http"
    WEBSOCKET = "websocket"


class Connection:
    """连接基类"""

    def __init__(self, config):
        self.config = config
        self.connected = False

    async def connect(self) -> None:
        raise NotImplementedError

    async def disconnect(self) -> None:
        raise NotImplementedError

    async def send(self, msg: Dict[str, Any]) -> None:
        raise NotImplementedError

    async def receive(self) -> Optional[Dict[str, Any]]:
        raise NotImplementedError


class HTTPConnection(Connection):
    """HTTP连接"""

    def __init__(self, config):
        super().__init__(config)
        self._session: Optional[aiohttp.ClientSession] = None
        self._base_url = f"http://{config.center_url}"
        self._poll_interval = 1.0
        self._message_queue: asyncio.Queue = asyncio.Queue()

    async def connect(self) -> None:
        self._session = aiohttp.ClientSession(
            timeout=aiohttp.ClientTimeout(total=self.config.timeout)
        )

        # 注册Agent
        register_url = f"{self._base_url}/api/v1/agents/register"
        async with self._session.post(register_url, json={
            "id": self.config.agent_id,
            "name": self.config.name,
            "type": "python",
        }) as resp:
            if resp.status == 200:
                self.connected = True
                logger.info(f"HTTP connection established: {self._base_url}")
                # 启动消息轮询
                asyncio.create_task(self._poll_messages())
            else:
                raise ConnectionError(f"Registration failed: {resp.status}")

    async def disconnect(self) -> None:
        self.connected = False
        if self._session:
            await self._session.close()
            self._session = None

    async def send(self, msg: Dict[str, Any]) -> None:
        if not self._session or not self.connected:
            raise ConnectionError("Not connected")

        msg_type = msg.get("type", "message")
        url = f"{self._base_url}/api/v1/{msg_type}"

        async with self._session.post(url, json=msg) as resp:
            if resp.status != 200:
                logger.error(f"Send failed: {resp.status}")

    async def receive(self) -> Optional[Dict[str, Any]]:
        try:
            msg = await asyncio.wait_for(
                self._message_queue.get(),
                timeout=0.1
            )
            return msg
        except asyncio.TimeoutError:
            return None

    async def _poll_messages(self) -> None:
        """轮询消息"""
        url = f"{self._base_url}/api/v1/agents/{self.config.agent_id}/messages"

        while self.connected:
            try:
                async with self._session.get(url) as resp:
                    if resp.status == 200:
                        messages = await resp.json()
                        for msg in messages:
                            await self._message_queue.put(msg)
                await asyncio.sleep(self._poll_interval)
            except Exception as e:
                logger.error(f"Poll error: {e}")
                await asyncio.sleep(self._poll_interval)


class WebSocketConnection(Connection):
    """WebSocket连接"""

    def __init__(self, config):
        super().__init__(config)
        self._ws: Optional[websockets.WebSocketClientProtocol] = None
        self._url = f"ws://{config.center_url}/ws"
        self._message_queue: asyncio.Queue = asyncio.Queue()

    async def connect(self) -> None:
        self._ws = await websockets.connect(
            self._url,
            ping_interval=30,
            ping_timeout=10
        )
        self.connected = True
        logger.info(f"WebSocket connection established: {self._url}")

        # 发送注册消息
        await self.send({
            "type": "register",
            "id": self.config.agent_id,
            "name": self.config.name,
        })

        # 启动消息接收
        asyncio.create_task(self._receive_loop())

    async def disconnect(self) -> None:
        self.connected = False
        if self._ws:
            await self._ws.close()
            self._ws = None

    async def send(self, msg: Dict[str, Any]) -> None:
        if not self._ws or not self.connected:
            raise ConnectionError("Not connected")

        await self._ws.send(json.dumps(msg))

    async def receive(self) -> Optional[Dict[str, Any]]:
        try:
            msg = await asyncio.wait_for(
                self._message_queue.get(),
                timeout=0.1
            )
            return msg
        except asyncio.TimeoutError:
            return None

    async def _receive_loop(self) -> None:
        """接收循环"""
        while self.connected and self._ws:
            try:
                data = await self._ws.recv()
                msg = json.loads(data)
                await self._message_queue.put(msg)
            except websockets.ConnectionClosed:
                self.connected = False
                logger.info("WebSocket closed")
                break
            except Exception as e:
                logger.error(f"Receive error: {e}")


class GRPCConnection(Connection):
    """gRPC连接（模拟实现）"""

    def __init__(self, config):
        super().__init__(config)
        self._channel: Optional[Any] = None
        self._stub: Optional[Any] = None
        self._message_queue: asyncio.Queue = asyncio.Queue()

    async def connect(self) -> None:
        # gRPC需要grpcio库
        try:
            import grpc
            from grpc import aio

            # 创建channel
            self._channel = aio.insecure_channel(self.config.center_url)
            self.connected = True
            logger.info(f"gRPC connection established: {self.config.center_url}")

        except ImportError:
            logger.warning("grpcio not installed, using HTTP fallback")
            # 回退到HTTP
            http_conn = HTTPConnection(self.config)
            await http_conn.connect()
            self._session = http_conn._session
            self.connected = True

    async def disconnect(self) -> None:
        self.connected = False
        if self._channel:
            await self._channel.close()

    async def send(self, msg: Dict[str, Any]) -> None:
        if not self.connected:
            raise ConnectionError("Not connected")

        # 模拟发送
        logger.debug(f"Sending: {msg}")

    async def receive(self) -> Optional[Dict[str, Any]]:
        try:
            msg = await asyncio.wait_for(
                self._message_queue.get(),
                timeout=0.1
            )
            return msg
        except asyncio.TimeoutError:
            return None